# Copy Number Variation Analysis Pipeline

## Introduction

The copy number variation (CNV) pipeline uses Affymetrix SNP 6.0 array data to infer genomic regions that are repeated and the copy number of these repeats. This pipeline is buildt on top of the existing TCGA level2 data generated by [Birdsuite] (https://www.broadinstitute.org/scientific-community/science/programs/medical-and-population-genetics/birdsuite/birdsuite), and uses the [DNAcopy](http://www.bioconductor.org/packages/release/bioc/html/DNAcopy.html) R-package to perform a circular binary segmentation (CBS) analysis [[1]](http://biostatistics.oxfordjournals.org/content/5/4/557.short) on TCGA samples, which measures copy-number on thousands of genomic targets simultaneously. First, CBS segments a chromosome into contiguous regions of equal copy number. Change-points, which signify a shift in copy number along a chromosome, are then estimated using array intensity values.   


## Data Processing Steps
The [Copy Number Liftover Workflow](Data_Dictionary/viewer/#?view=table-definition-view&id=copy_number_liftover_workflow) is implemented to convert the GRCh37 (hg19) probe set coordinates to the newer GRCh38 (hg38) genome build coordinates. GDC harmonization begins with the tangent.copynumber files that were originally distributed by TCGA. These files were generated by first normalizing array intensity values, estimating raw copy number, and performing tangent normalization, which subtracts variation that is found in a set of normal samples.   

Segmentation is performed on [Filtered Copy Number Segment](Data_Dictionary/viewer/#?view=table-definition-view&id=filtered_copy_number_segment) files using the DNAcopy R-package. This generates [Copy Number Segment](Data_Dictionary/viewer/#?view=table-definition-view&id=copy_number_segment) files, which associate chromosomal positions with the number of SNP6 probes associated with the segment and log2 ratio segment means (compared to the reference genome). Masked copy number segments are generated with the same method except that a filtering step is performed that removes segments associated with probe sets and regions that were previously indicated to have frequent germline copy-number variation.


| I/O | Entity | Format |
|---|---|---|
| Input | [Submitted Tangent Copy Number](Data_Dictionary/viewer/#?view=table-definition-view&id=submitted_tangent_copy_number) |  TXT |
| Output | [Copy Number Segment](Data_Dictionary/viewer/#?view=table-definition-view&id=copy_number_segment) or Masked Copy Number Segment | TXT  |

[1] Olshen, Adam B., E. S. Venkatraman, Robert Lucito, and Michael Wigler. "Circular binary segmentation for the analysis of array‚Äêbased DNA copy number data." Biostatistics 5, no. 4 (2004): 557-572.
